### PR TITLE
[Metricbeat] Simplify testing http Metricbeat modules

### PR DESCRIPTION
**Pull request by ruflin**
Currently most modules are tested against a docker container. This leads to long setup times and potentially flakyness. Also it requires additional setup to test actual changes on a module without running CI. The goal of this PR is to reduce this overhead, make it possible to easily test new data sets without having to write go code. Expected files were added to verify that changes had no effect on the generated data. The tests with the environment are still needed but should become less critical during development.

The structure and logic is inspired by the testing of the Filebeat modules. So far 3 metricsets were convert to test the implementation. It's all based on conventions:

* Tests outputs from a JSON endpoint must go int `_meta/testdata`
* A `testdata/config.yml` file must exists to specify url under which the testdata should be served
* A golden files is generated by adding `-expected.json`.

For a metricset to be converted it must have the reporter interface, be http and json based and only have 1 endpoint requested at the time. All metricsets should be converted to the reporter interface.

As there is now a more global view on the testing of a metricset, this code can potentially also take over the check to make sure that all fields are documented or at least the generated files can be used to do these checks.

To support metricsets which generate one or multiple events the output is always an array of JSON objects. These arrays can also contain errors, meaning also invalid data can be tested.

The `data.json` we had so far was hard to update and changed every time it was updated because it was pulled from a life instance. For the metricsets that are switched over to this testing, it's not the case anymore. The `data.json` is generated from the first event in the `docs.json`. This is by convention and allows to have a `docs.json` with a specially interesting event. This should make condition checks for which event should be shown also partially obsolete.

Future work:

* Support multiple endpoints: Elasticsearch metricsets do not work with the above model yet as they need multiple endpoints to be available at the same time. Config options for this could be introduced.
* Support more then .json: Currently only .json is supported. More config options could be added to support other data formats for example for the apache module
* Support other protocols then http: Not all modules are http based, 2-3 other comments protocols could be added.
* Extend with additional config options: Some metricsets need additional config options to be set for testing. It should be possible to pass these as part of the config.yml file.
* Generate the includes automatically: Currently if a new directory with testdata is added to a metricset, it will be discovered by the tests but then throws and error because the metricset is not registered. The metricset then has to be manually added to the `data_test.go` file. This works for now but potentially should be automated.

The overall goal of all the above is to have Metricbeat modules more and more config based instead of golang code based.

### CODE REVIEW

File: metricbeat/mb/testing/data_test.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+	// Read expected file
+	expected, err := ioutil.ReadFile(file + "-expected.json")
+	if err != nil {
+		t.Fatalf("could not read file: %s", err)
+	}
+
+	assert.Equal(t, string(expected), string(output))
+
+	if strings.HasSuffix(file, "docs.json") {
+		writeDataJSON(t, data[0], module, metricSetName)
```
Comment (jsoriano) [inline metricbeat/mb/testing/data_test.go]: It'd be nice to support here something to select events too, like #10367.

File: metricbeat/mb/testing/data_test.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+	}
+}
+
+func writeDataJSON(t *testing.T, data common.MapStr, module, metricSet string) {
+	// Add hardcoded timestamp
+	data.Put("@timestamp", "2019-03-01T08:05:34.853Z")
+	output, err := json.MarshalIndent(&data, "", "    ")
+	if err = ioutil.WriteFile("../../module/"+module+"/"+metricSet+"/_meta/data.json", output, 0644); err != nil {
```
Comment (jsoriano) [inline metricbeat/mb/testing/data_test.go]: Lots of similar paths around here, please add something to generate them (also thinking about x-pack).

File: metricbeat/module/traefik/health/_meta/testdata/docs.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (jsoriano) [inline metricbeat/module/traefik/health/_meta/testdata/docs.json]: There is a file like this one in `_meta/test/simple.json`, and there are similar files in other modules, not sure how they were generated, but should we use them instead? It can be weird to have both `test` and `testdata` directories with test files.

**Comment by cachedout** [comment]
`The tests with the environment are still needed but should become less critical during development.`

I take this to mean that the idea here is that the Docker environments would still be used in the CI as I (think?) they are now and that this new technique would just be something lighter and faster that a developer could use directly?

File: metricbeat/mb/testing/data_test.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+	}
+}
+
+func runTest(t *testing.T, file string, module, metricSetName, url string) {
+
+	// starts a server serving the given file under the given url
+	s := server(t, file, url)
+	defer s.Close()
+
+	metricSet := NewReportingMetricSetV2(t, getConfig(module, metricSetName, s.URL))
```
[the remark this answers is not in the document]
Comment (ruflin) [reply metricbeat/mb/testing/data_test.go]: Definitively a direction we can / should take this. I kept it simple to get started, we can always extend it.

File: metricbeat/mb/testing/data_test.go [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+func server(t *testing.T, path string, url string) *httptest.Server {
+
+	body, err := ioutil.ReadFile(path)
+	if err != nil {
+		t.Fatalf("could not read file: %s", err)
+	}
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path == url {
+			w.Header().Set("Content-Type", "application/json;")
```
[the remark this answers is not in the document]
Comment (ruflin) [reply metricbeat/mb/testing/data_test.go]: The other one is the `*.json` files discovery which needs adjustment. I did not make it more generic on purpose so we have something simple to get started and can iterate on it.

File: metricbeat/mb/testing/data_test.go [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+)
+
+type Config struct {
+	Type string
+	URL  string
+}
+
+func TestAll(t *testing.T) {
+
+	configFiles, _ := filepath.Glob("../../module/*/*/_meta/testdata/config.yml")
```
[the remark this answers is not in the document]
Comment (ruflin) [reply metricbeat/mb/testing/data_test.go]: Agree, we should add them as soon as we modify tests there.

File: metricbeat/mb/testing/data_test.go [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+	for _, f := range configFiles {
+		// get module and metricset name from path
+		s := strings.Split(f, "/")
```
[the remark this answers is not in the document]
Comment (ruflin) [reply metricbeat/mb/testing/data_test.go]: Good point, need to double check that.

File: metricbeat/mb/testing/data_test.go (hunk 1 above)
Comment (ruflin) [reply metricbeat/mb/testing/data_test.go]: I think we don't need this anymore because we have the docs.json convention which generates the data.json. So if we have a specific event that looks interesting, we should put it into the docs.json without any other events.

File: metricbeat/mb/testing/data_test.go (hunk 2 above)
Comment (ruflin) [reply metricbeat/mb/testing/data_test.go]: Abstracted it out a bit more but ignored x-pack for now. We should introduce it when we add a test for it.

Also introduced a constant for the expected extension.

File: metricbeat/module/traefik/health/_meta/testdata/docs.json (hunk 3 above)
Comment (ruflin) [reply metricbeat/module/traefik/health/_meta/testdata/docs.json]: I did not use the files in `test` in other directories on purpose to not have to deal with potentially other tests in this PR. In case we get decide to move forward with this approach, we should decide afterwards which existing tests should be removed. I kept all existing tests for now to make this PR independent. 

In the end I expect only `testdata` to be there but having both during a migration period.

**Comment by ruflin** [comment]
@cachedout Exactly. To test most cases a `.json` output from for example Elasticsearch 6.5.2 can be put into the directory and it's directly tested. This also allows us to test outputs from lots of different versions. The docker setup for development purpose would be more use to get these example json docs out.

Taking the Elasticsearch module as an example here, I expect us long term run the tests against many different versions to detect edge cases which might not be covered in the JSON files. These tests would not run as part of each PR but only on master or even a separate testing environment and report in case it finds some edge cases. Each module would then provide a tests matrix config against which versions things should be tested.

**Comment by ruflin** [comment]
@jsoriano The changes here definitively got heavily influenced by https://github.com/elastic/beats/blob/master/metricbeat/module/rabbitmq/mtest/server.go and I expect us to end up with something similar to that but in a more generic way. I initially started this PR with a more complex config for the tests but then threw it all away as I realised that we can probably cover 80% of the http module with a much simpler approach. What about the other 20%? Either we can extend the framework to fit, build something specific for these or keep it in go code. All of this fine as long all test frameworks stay reasonably simple.

In general I agree with all the abstraction requests you made above and I'm pretty sure at one stage we will need to implement them. But I prefer to implement them when we get to the point that we need them and have a test require it. So instead of having one massive PR I rather have 20 small ones improving the tests and get them in quickly. I hope that works for you.

File: metricbeat/mb/testing/data_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ruflin) [reply metricbeat/mb/testing/data_test.go]: I didn't have a Windows environment read to test this so I went the easy path and I'm using `filepath.ToSlash` for now.

**Comment by ruflin** [comment]
jenkins, test this
